add featuregate opt for kubeadm

fturib opened the pull request:
## What this PR does / why we need it:

allow option 'FeatureGates' when deploying kubeadm on the master's node of the cluster.
this option when set with "CoreDNS=true" allow to run CoreDNS as the default DNS server in the cluster instead of usual kube-dns

## Which issue this PR fixes 
Adds part of e2e integration requires for CoreDNS that is an ALPHA feature added to kubeadm 
See: kubernetes/kubeadm#446
it will need a second part in the project "test-infra"

## Special notes for your reviewer:
_In order to have the deployment of cluster run on GCE, I needed to fix issues on "terraform" tool usage (command line and config file). see gce/do and gce/gce.jsonnet_
**rollbacked in last commit. That was tied to the version of "terraform" used for testing.**

fturib commented:
I registered to CLA

fturib commented:
I fixed the line command of "kubeadm join" for nodes, so they can join the cluster. 
I think that should fix the global issues we have currently on the e2e testgrid.

fturib commented:
@luxas: can you review here ? or do you know who can be reviewer for this PR ?.
Thanks

fturib commented:
I hope one of view can comment and review this PR. Thanks!
/cc @shashidharatd @luxas @pipejakob

fturib commented on PROPOSAL.md:
```diff
@@ -43,3 +43,4 @@ How do we make this transparent? We can do this by using well documented and pop
 Taking kube-up as a conterexample (it's the antithesis of transparency): to understand the provisioning phase, a new user must trace through thousands of lines of imperative bash code. To understand the bootstrapping phase, a new user must read terribly complex salt configuration (look how many if branches are in this config [file](https://github.com/kubernetes/kubernetes/blob/master/cluster/salt ...
 
 The initial implementation should value transparency over production worthiness.
+
```
you are right. Was only to kick of another commit to have my CLA re-checked. Will removed.

fturib commented on phase1/gce/gce.jsonnet:
```diff
@@ -161,15 +161,18 @@ function(cfg)
             "k8s-kubeadm-kubernetes-version": "%(kubernetes_version)s" % p2,
             "k8s-kubeadm-kubelet-version": "%(kubelet_version)s" % p2,
             "k8s-kubeadm-enable-cloud-provider": (if std.objectHas(p2, "enable_cloud_provider") && p2.enable_cloud_provider then "true" else "false"),
+            "k8s-kubeadm-feature-gates": if std.objectHas(p2.kubeadm, "feature_gates") then "%(feature_gates)s" % p2.kubeadm else "",
```
adding ..

fturib commented on phase2/kubeadm/configure-vm-kubeadm-node.sh:
```diff
@@ -1,4 +1,4 @@
 # This is not meant to run on its own, but extends phase2/kubeadm/configure-vm-kubeadm.sh
 
-kubeadm join --token "$KUBEADM_TOKEN" "$KUBEADM_MASTER_IP:443" --skip-preflight-checks --discovery-token-unsafe-skip-ca-verification true
+kubeadm join --token "$KUBEADM_TOKEN" "$KUBEADM_MASTER_IP:443" --skip-preflight-checks --discovery-token-unsafe-skip-ca-verification
```
I see. Will remove

fturib commented on phase2/kubeadm/configure-vm-kubeadm-master.sh:
```diff
@@ -33,4 +33,24 @@ cloudProvider: "$CLOUD_PROVIDER"
+
+  for f in ${features[@]};
+  do
+    onef=(${f/=/ })
```
the whole game here is to go from a format : 
one string : " <key>=<value>,<key>=<value>,.."
multi lines:
"
featureGates:
   <key>: <value>
   <key>: <value>
   ...
"

I do in 2 steps: split in pairs by substitute , (comma) by " ' (space)
then on each pair : split by term by substitute = (equal) by " " (space)

not very pretty .. but works in bash.

fturib replied:
Will move the change of format to gce.jsonnet

fturib commented on phase1/gce/gce.jsonnet:
```diff
@@ -161,15 +161,18 @@ function(cfg)
           },
-          disk: [{
-            image: gce.os_image,
+          boot_disk: [{
```
yes, I noticed it. So I modified it too .. but I could not deploy with that second occurence as boot-disk. I guess that has to do that one is used directly for an "instance" and the other one is a "template".

Looking into the doc of gcloud create instance, 'disk' (as it was before) should be accepted, and there is no 'boot_disk' option. However on my test (with GCE) does work only adding the option boot_disk.

I will re-check that part.

fturib replied:
see here in terraform.io :

create instance : https://www.terraform.io/docs/providers/google/r/compute_instance.html
create instance template : https://www.terraform.io/docs/providers/google/r/compute_instance_template.html

first one work with "boot_disk" and second one with "disk".

fturib replied:
NOTE: I am not sure since when that option changed. "boot_disk" is now required. And I can validate that w/o that option the instance could not be created ...

but : how this configuration for GCE was working so far ? .... maybe recent change in the version of terraform.

fturib commented:
@shashidharatd: 
1- about bash - do you mean we CANNOT use parameter substitution ? (${..}). I also use the translation of string into an array. Is that also to be avoided ? What is the rule of what we can use or not ? 
What would you advice I use in this case ?

2- about Openstack and vSphere. The support of kubeadm is not implemented for those 2 environement. I am only adding an option for kubeadm.. so I can do it only for GCE.

=> I tested the code on GCE/kubeadm.

fturib commented:
I found a solution for the substitution, will make the change of format in jsonnet. That will be more compatible.

fturib commented on phase1/gce/gce.jsonnet:
```diff
@@ -15,6 +15,7 @@ function(cfg)
     node_firewall_rule: "%(cluster_name)s-node-all" % p1,
     release_bucket: "%s-kube-deploy-%s" % [gce.project, p1.cluster_name],
   };
+  local separate_fields(keyvalues) = std.join(" ", [std.join(" ", std.split(x,"=")) for x in std.split(keyvalues,",")]);
```
this function translate a parameter in format "key=value,key=value,.." into a set of values separated for spaces : "key value key value ..." which is a suitable format to be read as an array by bash script.

fturib commented on phase1/gce/gce.jsonnet:
```diff
@@ -157,11 +158,12 @@ function(cfg)
           } + if p2.provider == "kubeadm" then {
             "k8s-kubeadm-token": "${var.kubeadm_token}",
-            "k8s-kubeproxy-mode": "%(proxy_mode)s" % p2,
+            "k8s-kubeproxy-mode": if std.objectHas(p2, "proxy_mode") then "%(proxy_mode)s" % p2 else "",
```
proxy-mode was added before my commit. But I found an issue if test-infra and kubernetes-anywhere are not syncrhonized. Adding a 'compatibility' layer.

fturib commented:
I just uploaded a reviewed + squashed commit to take care of the comments:
- I backed some change of code on the format of terraform parameters - was due to the version of terraform to use (latest is 0.11.0, but we are using here the 0.9.4)
- moved the translation of format for feature-gates into the jsonnet script, in order to avoid the substitution on the bash script. Although I need to keep the array operations.
- added the KConfig parameter

NOTE: as said earlier, this feature is tied to kubeadm, which is currently implemented only for GCE. Thus I did not change anything for openstack of vsphere.

Thank you to review.!

fturib commented:
I update the stratup script for vm-master, in order to simplify the bash code (IMO).

fturib commented:
@shashidharatd : I think I addressed all your comments, would you be ok to resume the review ?Thanks!

fturib commented:
@pipejakob , @shashidharatd : I think I addressed all comments. Is it possible to go over the review for approval ? Thanks!

fturib commented:
@shashidharatd  : thanks for you review. Will address the changes for openstack - although I do not think I will be able to test those change ... And also the typo.

fturib commented:
@shashidharatd : I updated the PR with openstack support, and fix typo.

shashidharatd commented on phase1/gce/configure-vm.sh:
```diff
@@ -21,6 +21,7 @@ KUBEADM_ADVERTISE_ADDRESSES=$(get_metadata "k8s-kubeadm-advertise-addresses")
 KUBEADM_CNI_PLUGIN=$(get_metadata "k8s-kubeadm-cni-plugin")
 KUBEADM_MASTER_IP=$(get_metadata "k8s-kubeadm-master-ip")
 KUBEPROXY_MODE=$(get_metadata "k8s-kubeproxy-mode")
+KUBEADM_FEATURE_GATES=$(get_metadata "k8s-kubeadm-feature-gates")
```
we need to define this variable `KUBEADM_FEATURE_GATES` in openstack too.

fturib replied:
obviously yes ... will add.

shashidharatd commented on phase2/kubeadm/configure-vm-kubeadm-master.sh:
```diff
@@ -43,4 +42,25 @@ kubeProxy:
+#   key2: value2
+#   etc ...
+featuresdata=($KUBEADM_FEATURE_GATES)
+featureslength=(${#featuresdata[*]})
```
I am worried using `${` might not work for startup scripts. In openstack implementation we are using terraform templating which has special treatment for the patterns with `${`. Please refer https://www.terraform.io/docs/configuration/interpolation.html

Can we have some other alternative and avoid using `${` ?

fturib replied:
hum.. let me check what I can do based on this link.

shashidharatd commented on phase2/kubeadm/configure-vm-kubeadm-master.sh:
```diff
@@ -43,4 +42,25 @@ kubeProxy:
+  for ((ind=0; ind<$featureslength; ind++, ind++));
+  do
+    configfeatures="$configfeatures
+  ${featuresdata[ind]}: ${featuresdata[ind+1]}"
```
same as above comment

fturib replied:
My problem is the format of the parameter : at the end, in YAML file for kubeadmin, it is expected a multiline value. 
A way to avoid the loop in script and evaluation of expr, 
At the end I need: 

featureGates:
   <key1>: <value1>
   <key2>: <value2>
   ...

the list of key/values is provided by the initial parameter in jobs .. 

I do not like too much to prepare that format from jsonnet or terrarform  - that means pushed this parameter already YAML formatted in the cloud-init parameters (supposing I can provide here a multiline value without crazy encoding).

Best would be to have a tool at startup-script that could do the transformation (a simple function but not bash .. or a linux tool that can do csv-like to yaml).

Let me make some search here.

fturib commented:
@shashidharatd : I think I addressed your concerns
a- No expr evaluation like ${} in the shell scripts (replaced by `echo xxx | sed yyy`)
=> Thus I removed format transformation in jsonnet for the cloud-init variables
NOTE I verified it is working for gce/kubeadm environment with coredns option and without.
b- fixed openstack miss.

fturib commented:
Tank you @luxas  for the approval and merge !

luxas commented:
Sorry for the KubeCon-delay ;)
